rfc-117-container-runtime.md

```diff
@@ -0,0 +1,75 @@
+
+### Risks of this approach
+
+GOV.UK has not yet estimated the length of time this work will take, our current assumption is that it will take much less than 2 years to complete. There is a risk that the “unknown unknowns” GOV.UK discovers will prevent GOV.UK from finishing the containerisation work before support ends for Trusty. If this happens, GOV.UK may need to refocus our efforts to upgrade from Trusty in the existing infrastructure.
+
+GOV.UK infrastructure has been in a transitory state for several years and containerisation will prolong this state. In addition in the short term GOV.UK will be adding complexity making supporting the platform hard for RE-GOV.UK and 2nd line.
+
+The current migration (lift-and-shift to AWS) has been a complex and challenging piece of work. This has led to several engineers reporting feelings of burnout and exhaustion. If we fail to learn the lessons that GDS' past migration experiences have to offer there is a risk that doing this work could have significant negative consequences for the team, and for individuals' well-being. We should carefully consider how we structure teams, and how we can ensure that the teams have the time to do the work properly, get the training they need to learn new technologies, and have the agency they need to feel that they fully own their platform.
+
+Focusing on containerisation will mean fewer people will be available to work on improving other parts of the platform.
+
+GOV.UK has a significant amount of infrastructure outside of the applications themselves. In moving to a container based hosting platform, it’s possible that some existing behaviour may be missed, resulting in bugs.
+### Risks of not doing this
+
+Choosing not to move to a container runtime will mean that GOV.UK will have to upgrade from Ubuntu Trusty "in place" (i.e. by making significant changes to govuk-puppet). This will require significant effort for limited benefit, and is potentially wasted work as once the upgrade is done GOV.UK will still want to consider if running VMs is the right approach.
+
+GOV.UK will also have to tackle a number of difficult tooling upgrades (Puppet, in particular) if GOV.UK wants to return to a situation where GOV.UK does not use unsupported dependencies.
+
+GOV.UK will be harder for the rest of GDS to support, because its infrastructure will be significantly different. Almost all other service teams in GDS are using a container based hosting environment of some kind (mostly [CloudFoundry](https://www.cloudfoundry.org) or AWS ECS).
+### Follow up work
+
+GOV.UK will commit to doing these things:
+
+- Investigate the practicalities involved in using different container runtimes to host GOV.UK (for example, how difficult will it be to run a hybrid containerised / non-containerised environment in [GOV.UK PaaS](https://www.cloud.service.gov.uk) vs. say [AWS ECS](https://aws.amazon.com/ecs/))
```

Comment:
Given the proposal specifically talks about managed hosting providers, it might be clearer to stick to that language here, if that matches the intent?

```suggestion
- Investigate the practicalities involved in using different managed, container based hosting environments for GOV.UK (for example, how difficult will it be to run a hybrid containerised / non-containerised environment in [GOV.UK PaaS](https://www.cloud.service.gov.uk) vs. say [AWS ECS](https://aws.amazon.com/ecs/))
```

Reply:
Additionally, formatting it differently, and removing the names of different hosting providers might make it clearer:

- Investigate the practicalities involved in using different managed, container based hosting environments for GOV.UK
   - Including how difficult will it be to run a hybrid containerised / non-containerised environment

Reply:
Yep, addressed that in 9edd833